Mathlib/NumberTheory/NumberField/ClassNumber.lean

```diff
@@ -40,51 +41,103 @@ variable {K}
   · rw [← FractionalIdeal.absNorm_span_singleton (𝓞 K), Algebra.linearMap_apply,
       ← FractionalIdeal.coeIdeal_span_singleton, FractionalIdeal.coeIdeal_absNorm, hI, map_mul,
-      Nat.cast_mul, Rat.cast_mul, show Ideal.absNorm I₀ = Ideal.absNorm (I : Ideal (𝓞 K)) by rfl,
-      Rat.cast_natCast, Rat.cast_natCast, FractionalIdeal.coe_mk0,
-      FractionalIdeal.coeIdeal_absNorm, Rat.cast_natCast, mul_div_assoc, mul_assoc, mul_assoc]
-      at h_nm
+      cast_mul, Rat.cast_mul, absNorm_apply, Rat.cast_natCast, Rat.cast_natCast,
+      FractionalIdeal.coe_mk0, FractionalIdeal.coeIdeal_absNorm, Rat.cast_natCast, mul_div_assoc,
+      mul_assoc, mul_assoc] at h_nm
     refine le_of_mul_le_mul_of_pos_left h_nm ?_
-    exact Nat.cast_pos.mpr <| Nat.pos_of_ne_zero <| Ideal.absNorm_ne_zero_of_nonZeroDivisors J
+    exact cast_pos.mpr <| pos_of_ne_zero <| absNorm_ne_zero_of_nonZeroDivisors J
+
+theorem _root_.RingOfIntegers.isPrincipalIdealRing_of_isPrincipal_of_norm_le
+    (h : ∀ ⦃I : (Ideal (𝓞 K))⁰⦄, absNorm (I : Ideal (𝓞 K)) ≤ (4 / π) ^ nrComplexPlaces K *
+      ((finrank ℚ K)! / (finrank ℚ K) ^ (finrank ℚ K) * √|discr K|) →
+      Submodule.IsPrincipal (I : Ideal (𝓞 K))) :
+    IsPrincipalIdealRing (𝓞 K) := by
+  rw [← classNumber_eq_one_iff, classNumber, Fintype.card_eq_one_iff]
+  refine ⟨1, fun C ↦ ?_⟩
+  obtain ⟨I, rfl, hI⟩ := exists_ideal_in_class_of_norm_le C
+  simpa [← ClassGroup.mk0_eq_one_iff] using h hI
+
+theorem _root_.RingOfIntegers.isPrincipalIdealRing_of_isPrincipal_of_norm_le_of_isPrime
+    (h : ∀ ⦃I : (Ideal (𝓞 K))⁰⦄, (I : Ideal (𝓞 K)).IsPrime →
+      absNorm (I : Ideal (𝓞 K)) ≤ (4 / π) ^ nrComplexPlaces K *
+      ((finrank ℚ K)! / (finrank ℚ K) ^ (finrank ℚ K) * √|discr K|) →
+      Submodule.IsPrincipal (I : Ideal (𝓞 K))) :
+    IsPrincipalIdealRing (𝓞 K) := by
+  refine RingOfIntegers.isPrincipalIdealRing_of_isPrincipal_of_norm_le (fun I hI ↦ ?_)
+  rw [← mem_isPrincipalSubmonoid_iff,
+    ← prod_normalizedFactors_eq_self (nonZeroDivisors.coe_ne_zero I)]
+  refine Submonoid.multiset_prod_mem _ _ (fun J hJ ↦ mem_isPrincipalSubmonoid_iff.mp ?_)
+  by_cases hJ0 : J = 0
+  · simpa [hJ0] using bot_isPrincipal
+  rw [← Subtype.coe_mk J (mem_nonZeroDivisors_of_ne_zero hJ0)]
+  refine h (((mem_normalizedFactors_iff (nonZeroDivisors.coe_ne_zero I)).mp hJ).1) ?_
+  exact (cast_le.mpr <| le_of_dvd (absNorm_pos_of_nonZeroDivisors I) <|
+    absNorm_dvd_absNorm_of_le (le_of_dvd (UniqueFactorizationMonoid.dvd_of_mem_normalizedFactors
+    hJ))).trans hI
+
+theorem _root_.RingOfIntegers.isPrincipalIdealRing_of_isPrincipal_of_mem_primesOver
+    (h : ∀ ⦃p : ℕ⦄, p.Prime → p ≤ (4 / π) ^ nrComplexPlaces K *
+      ((finrank ℚ K)! / (finrank ℚ K) ^ (finrank ℚ K) * √|discr K|) →
+      ∀ (I : Ideal (𝓞 K)), I ∈ Ideal.primesOver (span {(p : ℤ)}) (𝓞 K) →
+      Submodule.IsPrincipal I) :
+    IsPrincipalIdealRing (𝓞 K) := by
+  refine RingOfIntegers.isPrincipalIdealRing_of_isPrincipal_of_norm_le_of_isPrime <|
+    fun P hP hPNorm ↦ ?_
```

Comment:
In most of the proof (except once) you work with `P.1`: can't you do a ` fun <P_a, P_b, hP>?

Reply:
Good idea.